### PR TITLE
Small optimisations - some faster array copies, reuse empty arrays

### PR DESCRIPTION
**Pull request by mkeskells**
take advantage of JVM optimisation when copying arrays to avoid zeroing

reuse empty arrays where it is simple to do so

### CODE REVIEW

File: src/library/scala/Array.scala [hunk 1]
```diff
@@ -51,15 +51,21 @@ class FallbackArrayBuilding {
...
+
+  val emptyBooleanArray = empty[Boolean]
+  val emptyByteArray    = empty[Byte]
+  val emptyCharArray    = empty[Char]
+  val emptyDoubleArray  = empty[Double]
+  val emptyFloatArray   = empty[Float]
+  val emptyIntArray     = empty[Int]
+  val emptyLongArray    = empty[Long]
+  val emptyShortArray   = empty[Short]
+  val emptyUnitArray    = empty[Unit]
```
Comment (lrytz) [inline src/library/scala/Array.scala]: this crashes
```
scala> java.lang.reflect.Array.newInstance(classOf[Unit], 0)
java.lang.IllegalArgumentException
  at java.lang.reflect.Array.newArray(Native Method)
  at java.lang.reflect.Array.newInstance(Array.java:75)
```

Comment (mkeskells) [reply src/library/scala/Array.scala]: thanks @lrytz - pushed a fix (I hope)

**Comment by mkeskells** [comment]
Mima issues 
```
static method emptyUnitArray()Array[scala.runtime.BoxedUnit] in class scala.Array does not have a correspondent in other version
filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("scala.Array.emptyUnitArray")
method emptyUnitArray()Array[scala.runtime.BoxedUnit] in object scala.Array does not have a correspondent in other version
filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("scala.Array.emptyUnitArray")
object scala.collection.mutable.WrappedArrayBuilder does not have a correspondent in other version
filter with: ProblemFilters.exclude[MissingClassProblem]("scala.collection.mutable.WrappedArrayBuilder$")
```

File: src/library/scala/collection/mutable/WrappedArrayBuilder.scala [hunk 2]
```diff
@@ -91,3 +112,24 @@ class WrappedArrayBuilder[A](tag: ClassTag[A]) extends ReusableBuilder[A, Wrappe
...
+          case java.lang.Double.TYPE    => new WrappedArray.ofDouble(Array.emptyDoubleArray)
+          case java.lang.Long.TYPE      => new WrappedArray.ofLong(Array.emptyLongArray)
+          case java.lang.Float.TYPE     => new WrappedArray.ofFloat(Array.emptyFloatArray)
+          case java.lang.Character.TYPE => new WrappedArray.ofChar(Array.emptyCharArray)
+          case java.lang.Byte.TYPE      => new WrappedArray.ofByte(Array.emptyByteArray)
+          case java.lang.Short.TYPE     => new WrappedArray.ofShort(Array.emptyShortArray)
+          case java.lang.Boolean.TYPE   => new WrappedArray.ofBoolean(Array.emptyBooleanArray)
+          case java.lang.Void.TYPE      => new WrappedArray.ofUnit(Array.emptyUnitArray)
+        }
+      else new WrappedArray.ofRef(Array.empty.asInstanceOf[Array[_ <: AnyRef]])
```
Comment (lrytz) [inline src/library/scala/collection/mutable/WrappedArrayBuilder.scala]: If I'm not mistaken, this is compiled as `Array.empty[Nothing].asInstanceOf[Array[_ <: AnyRef]]`? Seems to work, `Array.empty[Nothing].getClass` is `[LObject;`, but maybe it's cleaner to use `Array.empty[Object]`?

**Review by lrytz — COMMENTED** [review]
LGTM otherwise

File: src/library/scala/collection/mutable/WrappedArrayBuilder.scala (hunk 2 above)
Comment (mkeskells) [reply src/library/scala/collection/mutable/WrappedArrayBuilder.scala]: :-)  much simpler
updated to 
`else new WrappedArray.ofRef(Array.empty[AnyRef])`

**Review by lrytz — CHANGES_REQUESTED** [review]
Looking at this again made me think about the two new `ClassValue` caches. In a quick internal discussion, @retronym suggested to add new, lazily instantiated fields to `ClassTag` instead. It would work out well here because both of the new `ClassValue` caches are used in places where a `ClassTag` is available. I like this idea and think we should do it to avoid the two new caches.

**Comment by mkeskells** [comment]
> Looking at this again made me think about the two new `ClassValue` caches. In a quick internal discussion, @retronym suggested to add new, lazily instantiated fields to `ClassTag` instead. It would work out well here because both of the new `ClassValue` caches are used in places where a `ClassTag` is available. I like this idea and think we should do it to avoid the two new caches.

Added a commit with the changes.

**Comment by mkeskells** [comment]
MiMa issues (`emptyArray` and `emptyWrappedArray` are `private[scala]`)
```
    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.reflect.ClassTag#GenericClassTag.emptyArray"),
    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.reflect.ClassTag#GenericClassTag.emptyWrappedArray"),
    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.reflect.ManifestFactory#IntersectionTypeManifest.emptyArray"),
    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.reflect.ManifestFactory#IntersectionTypeManifest.emptyWrappedArray"),
    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.reflect.ManifestFactory#ClassTypeManifest.emptyArray"),
    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.reflect.ManifestFactory#ClassTypeManifest.emptyWrappedArray"),
    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.reflect.AnyValManifest.emptyArray"),
    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.reflect.AnyValManifest.emptyWrappedArray"),
    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.reflect.ClassTag.emptyArray"),
    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.reflect.ClassTag.emptyWrappedArray"),
    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.reflect.ManifestFactory#WildcardManifest.emptyArray"),
    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.reflect.ManifestFactory#WildcardManifest.emptyWrappedArray"),
    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.reflect.ManifestFactory#SingletonTypeManifest.emptyArray"),
    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.reflect.ManifestFactory#SingletonTypeManifest.emptyWrappedArray"),
    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.reflect.ClassManifestFactory#AbstractTypeClassManifest.emptyArray"),
    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.reflect.ClassManifestFactory#AbstractTypeClassManifest.emptyWrappedArray"),
    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.reflect.ManifestFactory#AbstractTypeManifest.emptyArray"),
    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.reflect.ManifestFactory#AbstractTypeManifest.emptyWrappedArray"),
    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.reflect.ClassTypeManifest.emptyArray"),
    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.reflect.ClassTypeManifest.emptyWrappedArray"),
```

**Review by lrytz — APPROVED** [review]
LGTM. I pushed a small cleanup for the mima filters.

In a follow-up, you could use the cached empty instance in `ClassTag.newArray` for primitives (you only do it for references):

```scala
scala> ClassTag.Int.newArray(0) eq ClassTag.Int.newArray(0)
res18: Boolean = false

scala> implicitly[ClassTag[String]].newArray(0) eq implicitly[ClassTag[String]].newArray(0)
res19: Boolean = true
```

**Comment by dwijnand** [comment]
Impacted Scala.js https://travis-ci.org/github/scala-js/scala-js-test-scala-nightly/jobs/723284403, chat on Gitter: https://gitter.im/scala/contributors?at=5f4f60db48237809375c2254 (intertwined with a chat about Shapeless)

**Comment by SethTisue** [comment]
#9335 fixes a regression here (in `Array.empty[Unit]`)

**Pull request merged**